### ISSUE TITLE
[🐛 Bug]: (Java)(Python) S3-S4 W3C Mode Not Supported Using ExpectedConditions

### ISSUE THREAD

**Issue by BJap** [SeleniumHQ/selenium#10759]
### What happened?

When the user attempts to wait for an `ExpectedCondition` to occur, some of them use an API that will result in a 404 error, which is to `findElement`.

This is due to the fact that unlike JWP, this command returns an error and ends the server session according to W3C spec:

https://www.w3.org/TR/webdriver/#find-element
https://www.w3.org/TR/webdriver/#dfn-no-such-element

vs JWP which returns the element or `null` if not found:

https://webdriver.io/docs/api/jsonwp/#findelement

This problem exists in both [Java](https://github.com/SeleniumHQ/selenium/blob/d594da2f5c2df8deeb56c9c45c67373edb75b6d5/java/src/org/openqa/selenium/support/ui/ExpectedConditions.java) and [Python](https://github.com/SeleniumHQ/selenium/blob/3824c3e9c5bb66b768e5ba2b56e8427e4f4f3c59/py/selenium/webdriver/support/expected_conditions.py) instances of `ExpectedConditions`/`expected_conditions`.

<HR>

Currently the problemed functions use:

```java
driver.findElement(locator);
```

```python
driver.find_element(*locator)
```

Both which can fail according to the W3C spec.

The solution would be to do something different such as leveraging `findElements`:

```java
driver.findElements(locator);
```

```python
driver.find_elements(*locator)
```

Source: https://www.w3.org/TR/webdriver/#dfn-find-elements

<HR>

This is because it will return the result no matter what, not an error, which could very well be an empty array. I feel the best solution would be to use that and grab the first element in most cases since this is the default behavior of the `findElement` function anyway. If there is no element, that would need to be handled as if the condition were true or false per function also.

"If result is empty, return [error](https://www.w3.org/TR/webdriver/#dfn-error) with [error code](https://www.w3.org/TR/webdriver/#dfn-error-code) [no such element](https://www.w3.org/TR/webdriver/#dfn-no-such-element). Otherwise, return the first element of result."

`visibilityOfAllElementsLocatedBy` is already using this function so it makes a good template even if it checks all elements (and we would just check the first one in most? cases).

<HR>

The affected functions from what I can see are:

**Java**

presenceOfElementLocated
visibilityOfElementLocated
textToBePresentInElementLocated
textToBePresentInElementValue
frameToBeAvailableAndSwitchToIt
invisibilityOfElementLocated
invisibilityOfElementWithText
elementSelectionStateToBe
attributeToBe
textToBe
textMatches
attributeContains
visibilityOfNestedElementsLocatedBy
presenceOfNestedElementLocatedBy
presenceOfNestedElementLocatedBy
presenceOfNestedElementsLocatedBy

**Python**

presence_of_element_located
visibility_of_element_located
text_to_be_present_in_element
text_to_be_present_in_element_value
text_to_be_present_in_element_attribute
frame_to_be_available_and_switch_to_it
invisibility_of_element_located
element_to_be_clickable
element_located_to_be_selected
element_located_selection_state_to_be
element_attribute_to_include

The unit tests for [ExpectedConditionsTest](https://github.com/SeleniumHQ/selenium/blob/d594da2f5c2df8deeb56c9c45c67373edb75b6d5/java/test/org/openqa/selenium/support/ui/ExpectedConditionsTest.java) and [expected_conditions_tests](https://github.com/SeleniumHQ/selenium/blob/trunk/py/test/selenium/webdriver/support/expected_conditions_tests.py) 'may' need looked at.

### How can we reproduce the issue?

Attempt to wait for any expected condition in the list above using a locator for an element that is not present in the DOM.

### Relevant log output

A server log will look something like:

```shell
[WD Proxy] Got response with status 404: {"value":{"error":"stale element reference","message":"stale element reference: element is not attached to the page document\n
```

A stack trace will look something like

```shell
at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
at org.openqa.selenium.remote.http.W3CHttpResponseCodec.createException(W3CHttpResponseCodec.java:187)
at org.openqa.selenium.remote.http.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:122)
at org.openqa.selenium.remote.http.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:49)
at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:158)
at io.appium.java_client.remote.AppiumCommandExecutor.execute(AppiumCommandExecutor.java:250)
at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:552)
at io.appium.java_client.DefaultGenericMobileDriver.execute(DefaultGenericMobileDriver.java:41)
at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:1)
at io.appium.java_client.android.AndroidDriver.execute(AndroidDriver.java:1)
at org.openqa.selenium.remote.RemoteWebDriver$RemoteWebDriverOptions$RemoteTimeouts.implicitlyWait(RemoteWebDriver.java:780)
```

### Operating System

Any

### Selenium version

Java and Python Selenium 3 and 4

### What are the browser(s) and version(s) where you see this issue?

Any using W3C mode

### What are the browser driver(s) and version(s) where you see this issue?

Any using W3C mode

### Are you using Selenium Grid?

No

**Comment by BJap** [comment]
One such example or question too is if such a change should consider all elements, or just maintain the behavior of before, and only consider the first element found. I imagine to avoid side effects it makes sense to change functions such as `invisibilityOfElementLocated` from behavior of:

```java
return !(driver.findElement(locator).isDisplayed());
```

to

```java
List<WebElement> elements = driver.findElements(locator);
return elements.size() == 0 || !(elements.get(0).isDisplayed());
```

instead of all checks like

```java
return driver.findElements(locator).stream().noneMatch(WebElement::isDisplayed);
```

Thoughts?

**Comment by jimevans** [comment]
@BJap I'm not sure I fully understand the issue. Is the issue that the expected conditions catch exceptions and ignore them until they're not thrown? What is the negative from the end user's perspective in the code they are using to call these expected conditions?

**Comment by AutomatedTester** [comment]
The JWP should be returning an error code of[ 7 for elements not found](https://www.selenium.dev/documentation/legacy/json_wire_protocol/#response-status-codes), not null.

For Element Stale Exception we could handle that in the expected exceptions but if you're trying to work with an element to see if it is displayed or what not, knowing that it has been switched out under you can be useful.

Could you add a failing test to help us visualise things better

**Comment by BJap** [comment]
> @BJap I'm not sure I fully understand the issue. Is the issue that the expected conditions catch exceptions and ignore them until they're not thrown? What is the negative from the end user's perspective in the code they are using to call these expected conditions?

In previous experience, the server would continue even if the element were not found, but now it will disconnect the session with the above mentioned error no matter if the exception is caught on the client side or not. The purpose of a wait for an element to change state like existence or visibility is to wait for such condition to occur, and then continue the test. The `WebDriverWait` uses some expected condition and when it returns true the test can continue.

For example, I could wait for an element to disappear, say one that identifies a form is on the screen. I click the save button and 'wait' for that form to disappear to indicate that the changes are saved (which isn't always immediate) and that I can perform actions now that it is done.


> The JWP should be returning an error code of[ 7 for elements not found](https://www.selenium.dev/documentation/legacy/json_wire_protocol/#response-status-codes), not null.
> 
> For Element Stale Exception we could handle that in the expected exceptions but if you're trying to work with an element to see if it is displayed or what not, knowing that it has been switched out under you can be useful.
> 
> Could you add a failing test to help us visualise things better

That actually seems to be correct now that I look at the code more and a 'true' is returned since the exception is swallowed.

In a nutshell, waiting for an element to disappear causes the server to shut down the session the moment it's gone, which makes a wait for an element to be gone to be useless.

In JWP the server would continue and let the client receive and exception and decide how to handle it, in W3C the server makes the decision in its current state and ends the session. The code in the links above shows how the client makes the choice to swallow the exceptions and fulfill the expected condition with a true rather than false.

We have also been following this issue in:
https://support.saucelabs.com/hc/en-us/requests/119971

The above link includes some tests runs and I can provide more as needed.

**Comment by diemol** [comment]
>In a nutshell, waiting for an element to disappear causes the server to shut down the session the moment it's gone, which makes a wait for an element to be gone to be useless.

Can you please share a Selenium WebDriver test that shows this behavior? 
The link you shared (https://support.saucelabs.com/hc/en-us/requests/119971) is not publicly accessible so we cannot see which test you are referring to.

**Comment by raolaks** [comment]
@diemol This is a Sauce Labs Real Device Test that can't be shared publicly. I can share the jWP and W3C log files here.
However, Sauce test execution video is internal only, but i can share that with you since you and i have access to it.

**Comment by diemol** [comment]
We discussed this internally with @raolaks and this will continue through Sauce Labs support channels. There is so far no evidence of an issue in Selenium. If that changes, we will reopen and handle it here.
